openHAB 4.2 release blog post (#475)
openhab/website: blog/2024-07-07-openhab-4-2-release.md

Huge thanks to all of them for keeping a high development pace for openHAB 4.2!

# Highlights

As you might have already noticed in the activity statistics, there was much development on openHAB - too much to list everything.
Please refer to the [**release notes**](https://github.com/openhab/openhab-distro/releases/tag/4.2.0) to check out what's new and noteworthy (e.g. breaking changes).

In the following sections, our maintainers and contributors are introducing some new features you don't want to miss!

## Notification Enhancements

@digitaldan: https://github.com/openhab/openhab-addons/issues/16934

Huge thanks to Dan Cunningham ([@digitaldan](https://github.com/digitaldan)) for doing all the work on openHAB Cloud, the Cloud Connector add-on and the iOS app!

Many thanks to our app maintainers [@weakfl](https://github.com/weakfl), Dan Cunningham ([@digitaldan](https://github.com/digitaldan)), Tim Bert ([@timbms](https://github.com/timbms)) [@mueller-ma](https://github.com/mueller-ma) and Danny Baumann ([@maniac103](https://github.com/maniac103)) for all their work on the apps!

## Core Runtime Enhancements

### Persistence Extensions

> @mherwege: https://github.com/openhab/openhab-core/pull/3736

Reviewer: Persistence Extensions have allowed the different scripting languages to easily persist states and retrieve persisted states or values calculated from these persisted states since a long time.

Since openHAB supports units of measure, persistence services can also return persisted values in a unit of measure.
However, the Persistence Extensions did not consider these units of measure and would return just a decimal value as the result of a calculation.
The Persistence Extension have been enhance to also return the unit of measure.
This allows rules and scripts to now just calculate with QuantityTypes without any implicit or explicit conversion to a DecimalType, while keeping the results consistent if different units of measure are used.
Note that this change may require adapting previous rules.

As persistence now allows forecasts of future states to be stored as well, the Persistence Extensions methods have been enhance to allow persisting future states and retrieving future values.
All persistence retrieval methods now have `last`, `next` and `between` variants of the methods, whereby `last` will look at the period from a time in the past until now, `next` will look from now to a future time and `between` will look between two provided times in past and/or future.

Persistence Extension methods now support working with TimeSeries instead of individual values when persisting or retrieving data.
TimeSeries are series of time/value pairs.
You can construct the series in a rule, e.g. from a TimeSeries retrieved from another persistence service or another item.

Also, there was no way before to retrieve the time of the last state change for an item from persistence.
The existing methods (`previousState()` or `previousState(true)`) would not give the right result if the same value had been persisted multiple times.
A new method `lastChange()` has been added to resolve this.

Scripting languages (rules DSL, javascript, jruby and Blockly) have been enhanced to fully support these new and enhanced Persistence Extensions.

For full documentation of the Persistence Extensions, have a look at [the documentation](https://openhab.org/docs/configuration/persistence.html#persistence-extensions-in-scripts-and-rules).
So much has changed that it is a whortwhile read to update your understanding of openHAB persistence extensions.

> @mherwege: https://github.com/openhab/openhab-core/pull/3736

Reviewer: Here is my proposed text. I won't have access to my PC starting tomorrow until after the release. So feel free to adjust:

Persistence Extensions have allowed the different scripting languages to easily persist states and retrieve persisted states or values calculated from these persisted states since a long time.

Since openHAB supports units of measure, persistence services can also return persisted values in a unit of measure.
However, the Persistence Extensions did not consider these units of measure and would return just a decimal value as the result of a calculation.
The Persistence Extension have been enhance to also return the unit of measure.
This allows rules and scripts to now just calculate with `QuantityTypes` without any implicit or explicit conversion to a `DecimalType`, while keeping the results consistent if different units of measure are used.
Note that this change may require adapting previous rules.

As persistence now allows forecasts of future states to be stored as well, the Persistence Extensions methods have been enhance to allow persisting future states and retrieving future values.
All persistence retrieval methods now have `last`, `next` and `between` variants of the methods, whereby `last` will look at the period from a time in the past until now, `next` will look from now to a future time and `between` will look between two provided times in past and/or future.

Persistence Extension methods now support working with `TimeSeries` instead of individual values when persisting or retrieving data.
TimeSeries are series of time/value pairs.
You can construct the series in a rule, e.g. from a `TimeSeries` retrieved from another persistence service or another item.

Also, there was no way before to retrieve the time of the last state change for an item from persistence.
The existing methods (`previousState()` or `previousState(true)`) would not give the right result if the same value had been persisted multiple times.
A new method `lastChange()` has been added to resolve this.

Scripting languages (rules DSL, javascript, jruby and Blockly) have been enhanced to fully support these new and enhanced Persistence Extensions.

For full documentation of the Persistence Extensions, have a look at [the documentation](https://openhab.org/docs/configuration/persistence.html#persistence-extensions-in-scripts-and-rules).
So much has changed that it is a whortwhile read to update your understanding of openHAB persistence extensions.

Author: Thanks for the text, I will update the already integrated one with a few rewordings.

This ensures we can always use the latest Webpack features when bundling the Main UI source code, and moves away from Node.js 16, which is now EOL.

Many thanks to Miguel Álvarez ([@GiviMAD](https://github.com/givimad)) for the Webpack upgrade!

### Semantic Model Templates

@justingeorgi: https://github.com/openhab/openhab-webui/pull/2105

### Persistence Settings Menu

Persistence configuration through the UI has been [supported since openHAB 4.0](/blog/2023-07-23-openhab-4-0-release.html#persistence-configuration), but unfortunately it was hidden in the add-on settings and difficult to find.

This finally changes with openHAB 4.2 with the introduction of the new persistence settings menu:

<p align="center"><img style="max-width: 70%;" src="/uploads/2024-07-07-openhab-4-2-release/persistence-menu.png"/></p>

The new page is reachable through the settings menu and allows to configure the default persistence service, persistence policies and installation of more persistence add-ons.

Thanks to Mark Herwege ([@mherwege](https://github.com/mherwege)) for the contribution!

### Item Form Enhancements

> @jimtng

Reviewer 2: Items are one of the core elements in an openHAB system.
To make viewing and editing them easier and more streamlined, the Item Form has received some minor cosmetic and functional improvements.
For dimensioned items, it is now easier to set the item's unit directly on the form with a default unit pre-filled.
Managing item's parent groups should now be easier too.

### Add-on Store Redesign

> @jimtng

Reviewer 2: The Add-on Store has been split up into sub-sections on the left navigation tree according to their category.
The main Add-on Store page now shows the list of all installed add-ons, grouped by category, for a quick overview.
The search box on the main Add-on Store page allows searching through all add-ons in every category.
The search box on the sub-category page searches only the selected category.

Browsing through the Add-ons Store and switching through the different sections is now much faster, thanks to cached results.
This makes browsing and installing add-ons through the UI easier and quicker.

Author: Thanks for the text, I have integrated it into the blog post and added a screenshot.

## Main UI Enhancements

_Florian Hotze ([@florian-h05](https://github.com/florian-h05)), openHAB Maintainer_

### Dependency & Build System Upgrades

openHAB 4.2 upgrades most of Main UI's dependencies to their latest version, including the charting engine [Apache ECharts](https://echarts.apache.org) and the date and time manipulation library [dayjs](https://day.js.org).
As Main UI is still using Vue 2, we would like to upgrade to Vue 3 in the future - by upgrading to the latest Vue 2.x version, we have taken a minor step towards Vue 3.

In addition to our runtime dependencies, we have also upgraded our build system to the latest tooling: Webpack 5 and Node.js 20.
This ensures we can always use the latest Webpack features when bundling the Main UI source code, and moves away from Node.js 16, which is now EOL.

Many thanks to Miguel Álvarez ([@GiviMAD](https://github.com/givimad)) for the Webpack upgrade!

### Semantic Model Templates

> @justingeorgi: https://github.com/openhab/openhab-webui/pull/2105

Author: Thanks, I have integrated it, will push the commit later.

The new Add-on Store sidebar menu and main page looks like this:

<p align="center"><img style="max-width: 70%;" src="/uploads/2024-07-07-openhab-4-2-release/add-on-store.png"/></p>

### Setup Wizard Extensions

With the introduction of [add-on suggestions](/blog/2023-12-22-openhab-4-1-release.html#add-on-suggestions) in openHAB 4.1, we made a first, big step towards a better first-user experience.

openHAB 4.2 continues this path by extending the setup wizard with a separate step for persistence.
In this step, suggested persistence add-ons are displayed together with a short explanation for the suggestion.
The user can then decide if he wants to install the suggested persistence add-ons or not, and install more add-ons in the next step:

<div align="center" class="row">
  <img style="max-width: 35%;" src="/uploads/2024-07-07-openhab-4-2-release/setup-wizard-persistence.png"/>
  <img style="max-width: 35%;" src="/uploads/2024-07-07-openhab-4-2-release/setup-wizard-addons.png"/>
</div>

Many thanks to Mark Herwege ([@mherwege](https://github.com/mherwege)), who already worked on the add-on suggestions in openHAB 4.1, for this enhancement!

### `oh-context` Component

> @justingeorgi: https://github.com/openhab/openhab-webui/pull/2533

Reviewer 3: ```suggestion
Advanced custom widgets have gotten more powerful with the new `oh-context` component.

This component allows widget authors to insert a code context block at any point in the widget tree and define three different types of objects available for use in expressions in all children of the `oh-context`:
- Variables: these context variables differ slightly from the normal widget variables in that they can have initial default values and also have complete communication with sub-widgets.
- Constants: context constants make for much more concise and readable expressions but unlike variable are only evaluated once when the widget loads.
- Functions: named context functions allow widget authors to efficiently reuse complex functions in multiple expressions.

Thanks to Justin Georgi ([@justingeorgi](https://github.com/JustinGeorgi)) for the contribution!
```

Author: Thanks for the text!

### Health UI

Ever heard of orphaned links before? Orphaned links are channel links, where either the Item or the channel (or possibly even both) are missing.

As you can imagine, orphaned links can and should be removed.
The new health UI allows to easily check if there are orphaned links and take care of them:

<p align="center"><img style="max-width: 70%;" src="/uploads/2024-07-07-openhab-4-2-release/health-menu.png"/></p>
<p align="center"><img style="max-width: 70%;" src="/uploads/2024-07-07-openhab-4-2-release/health-orphan-links.png"/></p>

The health UI is not only meant to display orphaned links, but also to provide a central place for health checks and maintenance tasks.
So stay tuned what gets added in the future!

Many thanks to Arne Seime ([@seime](https://github.com/seime)) for this addition!

### Blockly Enhancements

_Stefan Höhn ([@stefan-hoehn](https://github.com/stefan-hoehn)), Blockly Maintainer_

## Sitemap and Basic UI Enhancements

_Laurent Garnier ([@lolodomo](https://github.com/lolodomo)), openHAB Maintainer_

### Basic UI Enhancements

## Add-on Enhancements

> openHAB 4.2 brings X new add-ons, adding support for new devices:
> 
> - The [Amber Electric binding](/addons/bindings/amberelectric) provides access to the Australian energy retailer's Amber API with energy pricing and renewable energy data.
> - The [Basic Profiles transformation](/addons/transformations/basicprofiles), which provides a list of useful profiles and transformations.
> - The [Emotiva binding](/addons/bindings/emotiva) integrated Emotiva AV processors using the Emotiva Network Remote Control protocol.
> - The [Ephemeris binding](/addons/bindings/ephemeris) which provides access to openHAB Core's Ephemeris feature through Items.
> - The [French Government Energy Data binding](/addons/bindings/frenchgovtenergydata) provides regulated electricity prices in France.
> - The [IoTaWatt binding](/addons/bindings/iotawatt) integrated IoTaWatt Open WiFi Electric Power Monitor into openHAB.
> - The [MfFan binding](/addons/bindings/mffan) enables communication between openHAB and "Modern Forms" or "WAC Lighting" WIFI connected, smart, ceiling fans.
> - The [PegelOnline binding](/addons/bindings/pegelonline) allows to observe water levels from German rivers.
> - The [Salus binding](/addons/bindings/salus) facilitates seamless integration between openHAB and Salus Cloud.
> - The [SiemensHVAC binding](/addons/bindings/hvac) provides support for the Siemens HVAC controller ecosystem, and the Web Gateway interface OZW672.
> - The [TeslaPowerwall binding](/addons/bindings/teslapowerwall) enables the capture of key data from a Tesla Powerwall 2 into openHAB.
> - The [Whisper Speech-to-Text voice add-on](/addons/voice/whisperstt) provides a new offline voice recognition service for openHAB using OpenAI's Whisper model.

Reviewer 4: ```suggestion
openHAB 4.2 brings 22 new add-ons, adding support for new devices:

- The [AirGradient](/addons/bindings/airgradient) supports AirGradient air quality sensors.
- The [Amber Electric binding](/addons/bindings/amberelectric) provides access to the Australian energy retailer's Amber API with energy pricing and renewable energy data.
- The [ArgoClima binding](/addons/bindings/argoclima) for ArgoClima HVAC devices.
- The [Basic Profiles transformation](/addons/transformations/basicprofiles), which provides a list of useful profiles and transformations.
- The [CP750 binding](/addons/bindings/cp750) to communicate with a CP750, a common sound processor used in cinemas.
- The [Emotiva binding](/addons/bindings/emotiva) integrated Emotiva AV processors using the Emotiva Network Remote Control protocol.
- The [Ephemeris binding](/addons/bindings/ephemeris) which provides access to openHAB Core's Ephemeris feature through Items.
- The [Freeathome binding](/addons/bindings/freeathome) is a new binding for controlling free@home devices from ABB / Busch-Jaeger.
- The [Freecurrency binding](/addons/bindings/freecurrency) allows converting between different currencies and makes full use of the UoM currency capability.
- The [French Government Energy Data binding](/addons/bindings/frenchgovtenergydata) provides regulated electricity prices in France.
- The [GridBox binding](/addons/bindings/gridbox) for connecting to the Viessmann GridBox.
- The [Growatt binding](/addons/bindings/growatt) for Growatt solar inverters.
- The [IoTaWatt binding](/addons/bindings/iotawatt) integrated IoTaWatt Open WiFi Electric Power Monitor into openHAB.
- The [MfFan binding](/addons/bindings/mffan) enables communication between openHAB and "Modern Forms" or "WAC Lighting" WIFI connected, smart, ceiling fans.
- The [PanasonicBDP](/addons/bindings/panasonicbdp) for Panasonic Blu-ray players.
- The [PegelOnline binding](/addons/bindings/pegelonline) allows to observe water levels from German rivers.
- The [RadioBrowser binding](/addons/bindings/radiobrowser) to navigate and find internet radio broadcasts.
- The [Saicismart binding](/addons/bindings/saicismart) supports reading out vehicle data of cars running the iSMART system from SAIC.
- The [Salus binding](/addons/bindings/salus) facilitates seamless integration between openHAB and Salus Cloud.
- The [SiemensHVAC binding](/addons/bindings/hvac) provides support for the Siemens HVAC controller ecosystem, and the Web Gateway interface OZW672.
- The [Solar Forecast binding](/addons/bindings/solarforecast) to interface two different free photovoltaic forecast services.
- The [Sungrow binding](/addons/bindings/sungrow) is a Modbus binding that integrates Sungrow inverters.
- The [TeslaPowerwall binding](/addons/bindings/teslapowerwall) enables the capture of key data from a Tesla Powerwall 2 into openHAB.
- The [Whisper Speech-to-Text voice add-on](/addons/voice/whisperstt) provides a new offline voice recognition service for openHAB using OpenAI's Whisper model.
```

To allow storing multiple values to persistence in one go, Persistence Extensions methods now support working with Time Series instead of individual values when persisting or retrieving data.
Time Series are series of time/value pairs, and can be constructed in rules, e.g. by modifying a Time Series retrieved from another persistence service or another Item.

Also, there was no way before to retrieve the time of the last state change for an item from persistence.
The existing methods (`previousState()` or `previousState(true)`) would not give the right result if the same value had been persisted multiple times.
A new method `lastChange()` has been added to resolve this.

Scripting languages (Rules DSL, JavaScript, jRuby and Blockly) have implemented full support of these new and enhanced Persistence Extensions, please refer to the individual languages' documentation for more information.

As much has changed, it is worthwhile to read the [documentation](/docs/configuration/persistence.html#persistence-extensions-in-scripts-and-rules) to update your knowledge about openHAB persistence extensions and learn everything about the new features.

> 

Reviewer 4: A few words about Java21, hopefully not too technical....
Please feel free to change the test....

```suggestion

### Java 21

_Holger Friedrich ([@holgerfriedrich](https://github.com/holgerfriedrich)), openHAB maintainer_

Java 17 is required for all releases since openHAB 4.0, and Java 17 is still recommended for running openHAB.
But with version 4.2, we introduce Java 21 as an alternative.

We typically focus on Java releases with long-term support (LTS).
The next LTS release following after Java 17 is Java 21, released in 2023.
To allow for compiling openHAB with Java 21, we had to adapt our toolchain, upgrade dependencies, and also introduce some code changes.
This resulted in a comprehensive upgrade of almost all dependencies used in core. Just to name the most important ones: Karaf, Xtext/Xtend.
To allow for running openHAB with Java 21, a few more changes had to be introduced:
One notable difference between Java 17 and Java 21 is the change of the serialization format of DateTime in CLDR 42, which required some extra efforts in persistence to allow switching the Java version in an existing installation.

You can use either Java 17 or Java 21 to run your openHAB instance.
The openHAB distribution uses the Java 17 binary format which can run with both Java versions.
Please consider Java 21 as experimental, as it has not been tested thoroughly yet.
Note that Java 21 is only available for 64-bit systems.
If you use openHABian, you can install Java 21 using `openhabian-config`.
Please note that this is only possible if you installed the native 64-bit image.

Developers can even choose between three options: Java 17, Java 21 creating Java 17 binaries (default for OH if you use Java 21), and native Java 21 (binaries do not work on a Java 17 server).
Whereas we recommend Java 17 for running openHAB, development can be safely done on Java 21.
See the developer documentation for more information.
```

Author: Thanks, I have integrated this and just pushed to commit.

- The [RadioBrowser binding](/addons/bindings/radiobrowser) to navigate and find internet radio broadcasts.
- The [Saicismart binding](/addons/bindings/saicismart) supports reading out vehicle data of cars running the iSMART system from SAIC.
- The [Salus binding](/addons/bindings/salus) facilitates seamless integration between openHAB and Salus Cloud.
- The [SiemensHVAC binding](/addons/bindings/hvac) provides support for the Siemens HVAC controller ecosystem, and the Web Gateway interface OZW672.
- The [Solar Forecast binding](/addons/bindings/solarforecast) to interface two different free photovoltaic forecast services.
- The [Sungrow binding](/addons/bindings/sungrow) is a Modbus binding that integrates Sungrow inverters.
- The [TeslaPowerwall binding](/addons/bindings/teslapowerwall) enables the capture of key data from a Tesla Powerwall 2 into openHAB.
- The [Whisper Speech-to-Text voice add-on](/addons/voice/whisperstt) provides a new offline voice recognition service for openHAB using OpenAI's Whisper model.

But not only were new bindings added, some existing bindings also received massive improvements:

> @jlaur ???

Reviewer 5: Besides many small improvements here and there, nothing stands out in my memory, so I had a quick look at:
https://github.com/openhab/openhab-addons/pulls?q=is%3Apr+is%3Aclosed+milestone%3A4.2+label%3Aenhancement+

A few findings:
- BoschSHC: https://github.com/openhab/openhab-addons/pulls?q=is%3Apr+is%3Aclosed+milestone%3A4.2+label%3Aenhancement+in%3Atitle+boschshc, i.e. new devices supported: Door/Window Contact II, Universal Switch I+II, Smoke Detector II, Light/Shutter Control II, Dimmer, Smart Water Alarm
- HTTP: https://github.com/openhab/openhab-addons/pull/16282 (perhaps @J-N-K would like to mention something?)
- SamsungTV: https://github.com/openhab/openhab-addons/pull/11895

## openHABian Enhancements

> 

Reviewer 6: How does this look?

```suggestion
In openHABian, we've made a couple of under the hood enhancements to provide better support for the users running openHABian.
- Our latest image is now based on Bookworm the most recent release of Debian/Raspberry Pi OS
- Support for the new Raspberry Pi 5 has been added 
- 64-bit support is now official, check the docs for some caveats to running 64-bit vs 32-bit 

```

Author: Looks good, thanks!

- Holger ([@holgerfriedrich](https://github.com/holgerfriedrich)) for openHAB add-ons
- Florian ([@florian-h05](https://github.com/florian-h05)) for the openHAB Main UI
- Jimmy ([@jimtng](https://github.com/jimtng)) for the openHAB Main UI
- Laurent ([@lolodomo](https://github.com/lolodomo)) for the openHAB Basic UI

Many of our top committers are also maintainers, which means besides contributing code, they also review pull requests and participate in discussions on GitHub issues, which helps maintain our high level of quality across bindings, UIs and Core.
At present, we have 36 maintainers across all our repositories.
Huge thanks to all of them for keeping a high development pace for openHAB 4.2!

# Highlights

As you might have already noticed in the activity statistics, there was much development on openHAB - too much to list everything.
Please refer to the [**release notes**](https://github.com/openhab/openhab-distro/releases/tag/4.2.0) to check out what's new and noteworthy (e.g. breaking changes).

In the following sections, our maintainers and contributors are introducing some new features you don't want to miss!

> ## Notification Enhancements
> 
> @digitaldan: https://github.com/openhab/openhab-addons/issues/16934
> 
> Huge thanks to Dan Cunningham ([@digitaldan](https://github.com/digitaldan)) for doing all the work on openHAB Cloud, the Cloud Connector add-on and the iOS app!
> 
> Many thanks to our app maintainers [@weakfl](https://github.com/weakfl), Dan Cunningham ([@digitaldan](https://github.com/digitaldan)), Tim Bert ([@timbms](https://github.com/timbms)),
> [@mueller-ma](https://github.com/mueller-ma) and Danny Baumann ([@maniac103](https://github.com/maniac103)) for all their work on the apps!
> 

Reviewer 7: ```suggestion
## Mobile Notification Enhancements

_Dan Cunningham ([@digitaldan](https://github.com/digitaldan)), openHAB Maintainer_

With openHAB 4.2 we have updated our mobile push notifications architecture, bringing modern, cutting edge notification features to both our Android and iOS clients.

First, we have consolidated both mobile platforms to use Google Firebase Messaging, which greatly improves the reliability of push notifications from our cloud service as well as reduces the amount of work needed to add new features in a cross platform way.

This change has allowed us to add an incredible amount of new features to our app in a relatively short amount of time.

### Notification Display Enhancements

Notifications can now contain many more additional features, such as:

- Title
  - Replaces the default "openHAB" text with a user supplied version
- Media Attachments
  - Add images (or a short video on iOS) from an external URL or openHAB image item
- On Click Action
  - Navigate to a specific sitemap, page, popup widget, etc... in our UIs when the notification is clicked
- Action Buttons
  - Add up to 3 buttons that allow for executing commands, navigating through our UI's or open links in an embedded browser

<p align="center"><img style="max-width: 70%;" src="../tmp/notifications.png"/></p>

### Update / Collapse Notifications

We have also added some really useful ways to manage notifications already pushed to a device through rules and actions!

- Collapsible Notifications
  - Instead of seeing 100 messages that you left the garage door open, have a messages replace the existing shown one, so only the latest message is show on a user's device.
  - Update existing messages with new information, add additional actions, media, etc.. without showing multiple notifications.

<p align="center"><img style="max-width: 70%;" src="../tmp/collapse-notifications.png"/></p>

### Remove / Hide Notifications

Finally, this has been a long requested feature by users, so we are very excited to be able to support removable notifications in this release!

- Hide / Remove Notifications
  - Is an alert no longer valid? Need to display something only temporarily?
  - Using rules or actions, you can remove already pushed notifications from your devices.

### Scripting Language Support

These new notification enhancements are of course available across all our rule languages, but we have introduced additional helper APIs in our JS Scripting and Ruby script engine implementations to make this even easier to use.

### Upcoming iOS Changes

Its been a long time since we have pushed an official update to the iOS app. Unfortunately we have been unable to push updates do to a technical issue with our app store identifiers that Apple has not been able to resolve. After months and months of back and forth with their developer support team, we have decided to move forward with an alternative plan to release a brand new app.

This new app will still be called "openHAB" and will include all of the above features, more then a year's worth of bug fixes and improvements, the Apple Watch companion app and will continue to receive constant feature upgrades.
The new app will require at least iOS 16 and will drop openHAB 1.x support.

The existing app will continue to function, but be renamed to "openHAB V1". This "legacy" app will only receive security updates and critical fixes, but will support openHAB 1.x to at least 4.2 and will also work for devices running iOS 15 and earlier.
The Legacy app will not include Apple watch support nor support our new enhanced notifications features, although basic notifications will continue to be supported.  

Once both apps are available in the app store, the legacy app will prompt users to upgrade and make the switch

In the mean time, you can download our new app right now on [TestFlight](https://testflight.apple.com/join/0uFYONeF) and try out these amazing new features!

<a href="https://testflight.apple.com/join/0uFYONeF"><img src="https://developer.apple.com/assets/elements/icons/testflight/testflight-128x128_2x.png" width="50" alt="Download on TestFlight"></a>
Download the new openHAB app on [TestFlight](https://testflight.apple.com/join/0uFYONeF).

Many thanks to:

Florian Hotze ([@florian-h05](https://github.com/florian-h05)) for the enhanced JS-Scripting language support, docs, and code reviews of the binding.
Jimmy Tanagra ([@jimtng](https://github.com/jimtng)) for the enhanced Ruby language support

And also Many to our app maintainers [@mueller-ma](https://github.com/mueller-ma) and Danny Baumann ([@maniac103](https://github.com/maniac103)), [@weakfl](https://github.com/weakfl), Tim Bert ([@timbms](https://github.com/timbms)) for all their work on the mobile apps!
```

## Java 21 Support

_Holger Friedrich ([@holgerfriedrich](https://github.com/holgerfriedrich)), openHAB Maintainer_

Java 17 is required and still recommended for all openHAB 4.x releases.
But with version 4.2, we introduce Java 21 as an alternative:

We typically focus on Java releases with long-term support (LTS) and the next LTS release following after Java 17 is Java 21, released in 2023.
To allow compiling openHAB with Java 21, we had to adapt our toolchain, upgrade dependencies, and also introduce some code changes.
This resulted in a comprehensive upgrade of almost all dependencies used in core, including Karaf and Xtext/Xtend.
To allow for running openHAB with Java 21, a few more changes had to be introduced:
One notable difference between Java 17 and Java 21 is the change of the serialization format of DateTime in [CLDR 42](https://cldr.unicode.org/index/downloads/cldr-42), which required some extra efforts in persistence to allow switching the Java version in an existing installation.

> Now you can use both Java 17 and Java 21 to openHAB, but the openHAB distribution continues to use the Java 17 binary format which can run on both Java versions.

Reviewer 4: ```suggestion
Now you can use both Java 17 and Java 21 to run openHAB, but the openHAB distribution continues to use the Java 17 binary format which can run on both Java versions.
```

Please consider Java 21 support as experimental, as it has not been tested thoroughly yet.
Also note that Java 21 is only available for 64-bit systems.
If you use openHABian and have installed the 64-bit image, you can install Java 21 using `openhabian-config`.

Developers can even choose between three options:
Java 17, Java 21 creating Java 17 binaries (default for openHAB if you use Java 21), and native Java 21 (binaries will not work on a Java 17 server).
Whereas we recommend Java 17 for running openHAB, development can be safely done on Java 21.

> Read the [developer documentation](/docs/developer/#setup-the-development-environment) for more information.

Reviewer 4: ```suggestion
Please refer to the [developer documentation](/docs/developer/#setup-the-development-environment) for more information.
```

<p align="center"><img style="max-width: 70%;" src="/uploads/2024-07-07-openhab-4-2-release/notifications.png"/></p>

### Update / Collapse Notifications

We have also added some really useful ways to manage notifications already pushed to a device through rules and actions!

- Collapsible Notifications
  - Instead of seeing 100 messages that you left the garage door open, have a messages replace the existing shown one, so only the latest message is show on a user's device.
  - Update existing messages with new information, add additional actions, media, etc.. without showing multiple notifications.

<p align="center"><img style="max-width: 70%;" src="/uploads/2024-07-07-openhab-4-2-release/collapse-notifications.png"/></p>

### Remove / Hide Notifications

Finally, this has been a long requested feature by users, so we are very excited to be able to support removable notifications in this release!

- Hide / Remove Notifications
  - Is an alert no longer valid? Need to display something only temporarily?
  - Using rules or actions, you can remove already pushed notifications from your devices.

### Scripting Language Support

> These new notification enhancements are of course available across all our rule languages, but we have introduced additional helper APIs in our [JavaScript Scripting](/addons/automation/jsscripting/#cloud-notification-actions) and JRuby Scripting implementations to make this even easier to use.

Reviewer 2: ```suggestion
These new notification enhancements are of course available across all our rule languages, but we have introduced additional helper APIs in our [JavaScript Scripting](/addons/automation/jsscripting/#cloud-notification-actions) and [JRuby Scripting](https://openhab.github.io/openhab-jruby/main/OpenHAB/Core/Actions/Notification.html) implementations to make this even easier to use.
```

Author: Thanks for the link, I was already searching for a reference that I that provide yesterday, but thought I should recheck after I’ve seen a JRuby PR with the adjustments.

This change has allowed us to add an incredible amount of new features to our app in a relatively short amount of time.

### Notification Display Enhancements

Notifications now contain many more additional features, such as:

- Title
  - Replaces the default "openHAB" text with a user supplied version
- Media Attachments
  - Add images (or a short video on iOS) from an external URL or openHAB image item
- On Click Action
  - Navigate to a specific sitemap, page, popup widget, etc... in our UIs when the notification is clicked
- Action Buttons
  - Add up to 3 buttons that allow for executing commands, navigating through our UI's or open links in an embedded browser

<p align="center"><img style="max-width: 70%;" src="/uploads/2024-07-07-openhab-4-2-release/notifications.png"/></p>

### Update / Collapse Notifications

We have also added some really useful ways to manage notifications already pushed to a device through rules and actions!

- Collapsible Notifications

>   - Instead of seeing 100 messages that you left the garage door open, have a messages replace the existing shown one, so only the latest message is show on a user's device.
>   - Update existing messages with new information, add additional actions, media, etc.. without showing multiple notifications.

Reviewer 7: ```suggestion
  - Instead of seeing 100 messages that you left the garage door open, have a new message replace the existing one shown, so only the latest message is visible on a user's device.
  - Update existing messages with new information, add additional actions, updated media, etc.. without showing multiple notifications.
```

<p align="center"><img style="max-width: 70%;" src="/uploads/2024-07-07-openhab-4-2-release/collapse-notifications.png"/></p>

### Remove / Hide Notifications

Finally, this has been a long requested feature by users, so we are very excited to be able to support removable notifications in this release!

- Hide / Remove Notifications
  - Is an alert no longer valid? Need to display something only temporarily?
  - Using rules or actions, you can remove already pushed notifications from your devices.

### Scripting Language Support

These new notification enhancements are of course available across all our rule languages, but we have introduced additional helper APIs in our [JavaScript Scripting](/addons/automation/jsscripting/#cloud-notification-actions) and [JRuby Scripting](https://openhab.github.io/openhab-jruby/main/OpenHAB/Core/Actions/Notification.html) implementations to make this even easier to use.

> ### Upcoming iOS Changes

Reviewer 7: ```suggestion
### Android Client Availability

Keep on the lookout for an update to our Android client to support these awesome new features in the coming days and weeks.

<a href="https://play.google.com/store/apps/details?id=org.openhab.habdroid"><img src="https://play.google.com/intl/en_us/badges/images/generic/en_badge_web_generic.png" alt="Get it on Play Store" height="80"></a>

### iOS Client Availability
```

It's been a long time since we have pushed an official update to the iOS app.

> Unfortunately we have been unable to push updates do to a technical issue with our app store identifiers that Apple has not been able to resolve.
> After months and months of back and forth with their developer support team, we have decided to move forward with an alternative plan to release a brand-new app.

Reviewer 7: ```suggestion
Unfortunately we have been unable to push updates do to a technical issue with one of our app store identifiers that Apple has not been able to resolve.
After months of back and forth with Apple's developer support team, we have decided to move forward with an alternative plan to release a brand-new app.
```

> This new app will still be called "openHAB" and will include all of the above features, more than a year's worth of bug fixes and improvements, the Apple Watch companion app and will continue to receive constant feature upgrades.
> The new app will require at least iOS 16 and will drop openHAB 1.x support.

Reviewer 7: ```suggestion
Our new iOS app will still be called "openHAB" and will include all of great notification features mentioned above as well as more than a year's worth of bug fixes and improvements along with the Apple Watch companion app.
The new openHAB iOS app will require at least iOS 16 and will drop openHAB 1.x support.
```

The existing app will continue to function, but be renamed to "openHAB V1".
This "legacy" app will only receive security updates and critical fixes, but will support openHAB 1.x to at least 4.2 and will also work for devices running iOS 15 and earlier.
The legacy app will not include Apple Watch support nor support our new enhanced notifications features, although basic notifications will continue to be supported.

Once both apps are available in the app store, the legacy app will prompt users to upgrade and make the switch.

In the meantime, you can download our new app right now on [TestFlight](https://testflight.apple.com/join/0uFYONeF) and try out these amazing new features!

> <a href="https://testflight.apple.com/join/0uFYONeF"><img src="https://developer.apple.com/assets/elements/icons/testflight/testflight-128x128_2x.png" width="50" alt="Download on TestFlight"></a>

Reviewer 7: ```suggestion
<a href="https://testflight.apple.com/join/0uFYONeF"><img src="https://developer.apple.com/assets/elements/icons/testflight/testflight-128x128_2x.png" width="75" alt="Download on TestFlight"></a>
```

The health UI is not only meant to display orphaned links, but also to provide a central place for health checks and maintenance tasks.
So stay tuned what gets added in the future!

Thanks to Arne Seime ([@seime](https://github.com/seime)) for this nice addition!

### Blockly Enhancements

_Stefan Höhn ([@stefan-hoehn](https://github.com/stefan-hoehn)), Blockly Maintainer_

## Sitemap Enhancements

_Laurent Garnier ([@lolodomo](https://github.com/lolodomo)), openHAB Maintainer_

### Two possible behaviors for the Slider element

A new parameter named `releaseOnly` was added to the `Slider` element.
If `releaseOnly` parameter is specified, the new value is only sent on releasing the slider.
Otherwise values are sent while sliding which is now the default behavior.

Note: an implicit slider element (when `Default` element is used) will have the `releaseOnly` parameter set.
Its behavior is therefore to send the new value only when the slider is released.

### Press and release buttons

For a `Switch` element with mappings, each button can now behave either as a click button or as a press and release button.
The default behavior is a click button.

> To get a press and release behavior, provide two commands separated by a semicolon.

Reviewer 8: @lolodomo This sentence says semicolon, but the example below has a colon. Which one is correct?

Reviewer 9: Arghhh !
You're right, good catch , I made the confusion in the docs first, due to my limited English.
I will prepare a doc fix.

Reviewer 9: https://github.com/openhab/openhab-docs/pull/2329

- The [RadioBrowser binding](/addons/bindings/radiobrowser) to navigate and find internet radio broadcasts.
- The [Saicismart binding](/addons/bindings/saicismart) supports reading out vehicle data of cars running the iSMART system from SAIC.
- The [Salus binding](/addons/bindings/salus) facilitates seamless integration between openHAB and Salus Cloud.
- The [SiemensHVAC binding](/addons/bindings/hvac) provides support for the Siemens HVAC controller ecosystem, and the Web Gateway interface OZW672.
- The [Solar Forecast binding](/addons/bindings/solarforecast) to interface two different free photovoltaic forecast services.
- The [Sungrow binding](/addons/bindings/sungrow) is a Modbus binding that integrates Sungrow inverters.
- The [TeslaPowerwall binding](/addons/bindings/teslapowerwall) enables the capture of key data from a Tesla Powerwall 2 into openHAB.
- The [Whisper Speech-to-Text voice add-on](/addons/voice/whisperstt) provides a new offline voice recognition service for openHAB using OpenAI's Whisper model.

But not only were new bindings added, some existing bindings also received massive improvements:

> 
> @jlaur ???

Reviewer 5: ```suggestion
- The [Bosch Smart Home binding](/addons/bindings/boschshc) has added support for new devices: Door/Window Contact II, Universal Switch I+II, Smoke Detector II, Light/Shutter Control II, Dimmer and Smart Water Alarm.
- The [SamsungTV binding](/addons/bindings/samsungtv) has extended support for Frame TV with new channels as well as other improvements.
- The [Shelly binding](/addons/bindings/shelly) has added support for new devices and Range Extender.
- The HTTP binding](/addons/bindings/http) has received an overhaul with new configuration options, channels and bug fixes.
```

Reviewer 5: @openhab/add-ons-maintainers - anything more to add here?

Reviewer 9: You could also mention stabilization for netatmo binding with many PRs merged.

Reviewer 5: Updated above.

Also note that Java 21 is only available for 64-bit systems.
If you use openHABian and have installed the 64-bit image, you can install Java 21 using `openhabian-config`.

Developers can even choose between three options:
Java 17, Java 21 creating Java 17 binaries (default for openHAB if you use Java 21), and native Java 21 (binaries will not work on a Java 17 server).
Whereas we recommend Java 17 for running openHAB, development can be safely done on Java 21.

Please refer to the [developer documentation](/docs/developer/#setup-the-development-environment) for more information.

## Core Runtime Enhancements

### Persistence Extensions

_Mark Herwege ([@mherwege](https://github.com/mherwege)), openHAB Contributor_

Persistence Extensions have allowed the different scripting languages to easily persist states and retrieve persisted states or values calculated from these persisted states for a long time.

As openHAB supports units of measurument (UoM), persistence services can also return persisted values with a unit of measurement.
However, the Persistence Extensions did not consider these units of measurement and returned just a decimal value as the result of their calculations.

> With openHAB 4.2, Persistence Extension have been enhanced to also return the unit of measurement.

Reviewer 5: ```suggestion
With openHAB 4.2, Persistence Extensions have been enhanced to also return the unit of measurement.
```

### Persistence Extensions

_Mark Herwege ([@mherwege](https://github.com/mherwege)), openHAB Contributor_

Persistence Extensions have allowed the different scripting languages to easily persist states and retrieve persisted states or values calculated from these persisted states for a long time.

As openHAB supports units of measure, persistence services can also return persisted values in a unit of measure.
However, the Persistence Extensions did not consider these units of measure and would return just a decimal value as the result of a calculation.

> With openHAB 4.2, Persistence Extension have been enhance to also return the unit of measure.

Reviewer: have been enhanced

openHAB 4.2 extends the Persistence Extensions to allow persisting future states and retrieving future values.
All persistence retrieval methods now have _last_, _next_ and _between_ variants of the methods, whereby _last_ will use the period from a time in the past until now, _next_ will use the period from now to a future time and _between_ will use the period between two provided points in time in the past and/or future.

To allow storing multiple values to persistence in one go, Persistence Extensions methods now support working with TimeSeries instead of individual values when persisting or retrieving data.
Time Series are series of time/value pairs, and can be constructed in rules, e.g. by modifying a Time Series retrieved from another persistence service or another Item.

Also, there was no way before to retrieve the time of the last state change for an item from persistence.
The existing methods (`previousState()` or `previousState(true)`) would not give the right result if the same value had been persisted multiple times.
A new method `lastChange()` has been added to resolve this.

> Scripting languages (Rules DSL, JavaScript, jRuby and Blockly) have implement full support of these new and enhanced Persistence Extensions, please refer to the individual languages' documentation for more information.

Reviewer: small change: have implemented
